spark3-extensions/src/main/scala/org/apache/spark/sql/catalyst/optimizer/RewriteMergeInto.scala

```diff
@@ -0,0 +1,204 @@
+        val writeInfo = newWriteInfo(target.schema)
+        val mergeBuilder = target.table.asMergeable.newMergeBuilder("delete", writeInfo)
+        val targetTableScan =  buildScanPlan(spark, target.table,
+          target.output, mergeBuilder, prunedTargetPlan, isCountCheckEnabled(target.table, actions))
+
+        // Construct an outer join to help track changes in source and target.
+        // TODO : Optimize this to use LEFT ANTI or RIGHT OUTER when applicable.
+        val sourceTableProj = source.output ++ Seq(Alias(lit(true).expr, ROW_FROM_SOURCE)())
+        val targetTableProj = target.output ++ Seq(Alias(lit(true).expr, ROW_FROM_TARGET)())
+        val newTargetTableScan = Project(targetTableProj, targetTableScan)
+        val newSourceTableScan = Project(sourceTableProj, source)
+        val joinPlan = Join(newSourceTableScan, newTargetTableScan, FullOuter, Some(cond), JoinHint.NONE)
+
+        // Construct the plan to replace the data based on the output of `MergeInto`
+        val mergeParams = MergeIntoParams(
+          isSourceRowNotPresent = IsNull(findOutputAttr(joinPlan, ROW_FROM_SOURCE)),
+          isTargetRowNotPresent = IsNull(findOutputAttr(joinPlan, ROW_FROM_TARGET)),
+          matchedConditions = actions.map(getClauseCondition),
+          matchedOutputs = actions.map(actionOutput(_, targetOutputCols)),
+          notMatchedConditions = notActions.map(getClauseCondition),
+          notMatchedOutputs = notActions.map(actionOutput(_, targetOutputCols)),
+          targetOutput = targetOutputCols :+ Literal(false),
+          deleteOutput = targetOutputCols :+ Literal(true),
+          joinedAttributes = joinPlan.output
+        )
+        val joinKeysFromTarget = targetOutputCols.filter (
+          attr => cond.references.exists(attr.semanticEquals(_))
+        )
+        val writePlan = buildWritePlan(joinPlan, Project(target.output, joinPlan), target.table, joinKeysFromTarget)
+        val mergePlan = MergeInto(mergeParams, target, writePlan)
+        val batchWrite = mergeBuilder.asWriteBuilder.buildForBatch()
+        ReplaceData(target, batchWrite, mergePlan)
+    }
+  }
+
+  def buildWritePlan(childPlan: LogicalPlan,
+                     planToResolveFrom: LogicalPlan,
+                     table: Table, targetJoinAttrs: Seq[Attribute]): LogicalPlan = {
+    val iceTable = table.asInstanceOf[SparkTable].table()
+    val globalSortEnabled = isGlobalSortEnabled(table)
+    val partitionExpressions = toCatalyst(toRequiredDistribution(iceTable.spec), planToResolveFrom)
+    val sortExpressions: Seq[SortOrder] =  toCatalyst(toRequiredDistribution(iceTable.spec(),
+      iceTable.sortOrder(), false), planToResolveFrom).asInstanceOf[Seq[SortOrder]]
+    val dist = toRequiredDistribution(iceTable.spec(),
+      iceTable.sortOrder(), true)
+    val globalSortExprs: Seq[SortOrder] = toCatalyst(dist, planToResolveFrom).asInstanceOf[Seq[SortOrder]]
```

Comment:
I find this method really hard to follow. This creates expressions for every possible combination and then decides which ones to use.

The logic below also seems more complicated than it needs to be, I think because it uses a different order to check the possible combinations of `isGlobal`, `isPartitioned` and `isSorted`. For example, if `isGlobal` is true, but there is no ordering, then it needs to create an order rather than just not ordering. I was expecting something a bit simpler, like this:

```scala
val hasSortOrder = !table.sortOrder.isUnordered();
if (hasSortOrder) {
  if (distributionMode == Sort) {
    val order = toCatalyst(toOrderedDistribution(spec, order, true))
    val roundRobin = Repartition(numShufflePartitions, shuffle = true, childPlan)
    Sort(order, global = true, roundRobin)
  } else if (distributionMode == Partition) {
    val order = toCatalyst(order)
    val hashPartitioned = RepartitionByExpression(toCatalyst(toClusteredDistribution(spec)), childPlan, numShufflePartitions)
    Sort(order, global = false, hashPartitioned)
  } else {
    val order = toCatalyst(order)
    Sort(order, global = false, childPlan)
  }
} else if (distributionMode == Partition) {
  RepartitionByExpression(toCatalyst(toClusteredDistribution(spec)), childPlan, numShufflePartitions)
} else {
  childPlan
}
```

That is a bit simpler because global ordering is only considered if the table has an ordering. It may still be a good idea to do what you're doing and create an order from join attributes, but I'd like to make sure that we're choosing to do that. The simplest option is to turn off sorting if there is no global order.

Another alternative if the "distribution mode" is to sort: we could sort by partition keys, _and_ the join attributes. As long as we are using a range partitioner, we should try to minimize files created in each partition.

Also: in my code above, I added a round-robin shuffle before the range partitioner to avoid running the join and merge twice (once to calculate skew and once to produce records).

Reply:
Following up on this, I think we should implement the proposal from our [discussion about Flink hash distribution](https://github.com/apache/iceberg/pull/2064#discussion_r559018257). That PR is going to add `write.distribution-mode` with 3 values: `none`, `hash`, and `range`. Here's what we would use them for here:

| Spark | `none` | `hash` | `range` |
|-|--------|-------------|--------|
| unordered | no distribution, locally sort by partition key | hash distribute by partition key, locally sort by partition key | range distribute by partition key, locally sort by partition key |
| ordered | no distribution, locally sorted | hash distribute by partition key, locally sorted | globally sorted |

Or in terms of the methods you've added:

| Spark | `none` | `partition` | `sort` |
|-|--------|-------------|--------|
| unordered | no distribution, local sort by `toOrderedDistribution(spec, order, true)` | distribute by `toClusteredDistribution(spec)`, local sort by `toOrderedDistribution(spec, order, true)` | global sort by `toOrderedDistribution(spec, order, true)` |
| ordered | no distribution, local sort by `toOrderedDistribution(spec, order, true)` | distribute by `toClusteredDistribution(spec)`, local sort by `toOrderedDistribution(spec, order, true)` | global sort by `toOrderedDistribution(spec, order, true)` |

The result of `toOrderedDistribution(spec, order, true)` is always used to sort, whether locally or globally. If the sort order is unordered, then it will infer the order from the partition spec just as we wanted from the top table. For hash partitioning, we can use `toClusteredDistribution` that you added. Then the only other concern is that we need to add a round-robin shuffle before the global sort.

With that addition, let me revise my code from above:

```scala
val distributionMode = table.properties.getOrDefault("write.distribution-mode", "global")
val order = toCatalyst(toOrderedDistribution(spec, order, true))
distributionMode.toLower(Locale.ROOT) match {
  case "none" =>
    Sort(order, global = false, childPlan)
  case "hash" =>
    val clustering = toCatalyst(toClusteredDistribution(spec))
    val hashPartitioned = RepartitionByExpression(clustering, childPlan, numShufflePartitions)
    Sort(order, global = false, hashPartitioned)
  case "range" =>
    val roundRobin = Repartition(numShufflePartitions, shuffle = true, childPlan)
    Sort(order, global = true, roundRobin)
}
```

Reply:
@rdblue Thank you very much !!

Reply:
+1 here